### PR TITLE
Current work on improving option parser

### PR DESCRIPTION
**Pull request by joelmccracken**
not actually for pull request

### CODE REVIEW

**Comment by zzak** [comment]
@joelmccracken I just reviewed your docs here and have a few comments:
- The "generating help" example is good, I'd like to add this!
- We should find a way to break up the "Complete example" and still keep it complete.

What I mean is, the complete example is good because it shows you how to implement `optparse` in your own class. I'd like to still do this, while breaking it up into more manageable chunks that are well explained (and still make sense as a whole). Perhaps we could do this with composition?

**Comment by zzak** [comment]
@joelmccracken I want to commit the "generating help" example, I started to rewrite some of it and came up with these 2 versions (one using a `class`, the other using `Class.new`):

https://gist.github.com/zzak/7027206

What do you think?

**Comment by joelmccracken** [comment]
That works. Thanks!

I found the wrapping around the OptionParser instance confusing at first. My rationale f

**Comment by joelmccracken** [comment]
(Sorry about that). My rationale for removing the whole wrapper component was just to simplify the relevant OptionParser code. 

So sure. Delegation would work. When I get some time I will try to break it up that way. 

Thanks for your help!

**Comment by joelmccracken** [comment]
Out of the two you linked to above, I think I prefer the plain `class` version. I'm going to add it to this PR.

**Comment by joelmccracken** [comment]
I really don't understand how the code that supports bash completion is supposed to work or be integrated with the shell. Any thoughts?

**Comment by joelmccracken** [comment]
@zzak so I tried using a few forms of delegation, but I think it just generally obscured the complete example more than it served to clarify.

Instead, I separated the code for each option into its own private method. I also tried to unify around some of the language in "http://optik.sourceforge.net/doc/1.4/tao.html".

WDYT?

**Comment by joelmccracken** [comment]
Is there something I can do to help move this along, like building a patch? Maybe submitting a ticket on the ruby issue tracker? 

<3

**Comment by zzak** [comment]
@joelmccracken Sorry for not getting back to you sooner, I'll look into this tonight. <3 <3 <3

**Comment by joelmccracken** [comment]
No man, its ok! I'm not trying to bug you or anything.

One of the general things I don't like in life is the principle of helping
someone, but making work for them in the process.

Like, giving someone flowers is a really nice gesture. The thing is, when
you give flowers, you are also giving them the chore to go find a vase to
put them in, cut the stems, and add water. Its like, "here are some
flowers, now go spend 10 mins doing a chore I've just sprung on you".
Randomly giving someone a puppy would be similar, but much worse, which is
why I think people don't actually give each other puppies and kittens.

I guess what I'm saying is PRs are nice, but are like giving someone
flowers. It comes with some random work attached.

**Comment by zzak** [comment]
I appreciate your concern, but this is apart of OSS.

Thank you for your patience and contribution to the project, without contributors like yourself this project wouldn’t be possible.

**Comment by zzak** [comment]
@joelmccracken I've committed the first part of the patch for "Generating help".

Could you rebase your branch so that its easier to review the rest of the patch?

**Comment by joelmccracken** [comment]
@zzak I rebased. Thanks for making the patch!

File: lib/optparse.rb [hunk 1]
```diff
@@ -129,111 +129,147 @@
 #   require 'optparse/time'
 #   require 'ostruct'
 #   require 'pp'
-#
```
Comment (zzak) [inline lib/optparse.rb:132]: I'm ok with this extra line to separate requires from class.

File: lib/optparse.rb [hunk 2]
```diff
@@ -129,111 +129,147 @@
...
 #     CODE_ALIASES = { "jis" => "iso-2022-jp", "sjis" => "shift_jis" }
 #
 #     #
 #     # Return a structure describing the options.
 #     #
-#     def self.parse(args)
-#       # The options specified on the command line will be collected in *options*.
-#       # We set default values here.
+#     def parse(args)
+#
```
Comment (zzak) [inline lib/optparse.rb]: This one could go :scissors:

File: lib/optparse.rb [hunk 3]
```diff
@@ -129,111 +129,147 @@
...
 #   require 'ostruct'
 #   require 'pp'
-#
 #   class OptparseExample
-#
 #     CODES = %w[iso-2022-jp shift_jis euc-jp utf8 binary]
 #     CODE_ALIASES = { "jis" => "iso-2022-jp", "sjis" => "shift_jis" }
 #
 #     #
 #     # Return a structure describing the options.
```
Comment (zzak) [inline lib/optparse.rb:262]: Perhaps we could give this a better explanation?

File: lib/optparse.rb [hunk 4]
```diff
@@ -129,111 +129,147 @@
...
 #     #
-#     def self.parse(args)
-#       # The options specified on the command line will be collected in *options*.
-#       # We set default values here.
+#     def parse(args)
+#
+#       # The options specified on the command line will be collected in
+#       # *options*.
 #       options = OpenStruct.new
+#       # We set default values here.
```
Comment (zzak) [inline lib/optparse.rb]: I wonder if these belong in an initializer?

File: lib/optparse.rb [hunk 5]
```diff
@@ -129,111 +129,147 @@
...
 #
-#         opts.separator ""
-#         opts.separator "Specific options:"
 #
-#         # Mandatory argument.
-#         opts.on("-r", "--require LIBRARY",
-#                 "Require the LIBRARY before executing your script") do |lib|
-#           options.library << lib
-#         end
+#     private
```
Comment (zzak) [inline lib/optparse.rb]: I think the parser should be public interface, also maybe we should cache it so we're not regenerating every call to `option_parser`, ie: `@parser ||= #...`

File: lib/optparse.rb [hunk 6]
```diff
@@ -129,111 +129,147 @@
...
+#       OptionParser.new do |parser|
+#         parser.banner = "Usage: example.rb [options]"
+#         parser.separator ""
+#         parser.separator "Specific options:"
 #
-#         # Cast 'delay' argument to a Float.
-#         opts.on("--delay N", Float, "Delay N seconds before executing") do |n|
-#           options.delay = n
-#         end
+#         # add additional options
```
Comment (zzak) [inline lib/optparse.rb:281]: I'm not sure having all these additional options adds any value to the example without a proper usecase.

**Comment by zzak** [comment]
Perhaps this is beyond the scope of the example.. but since we're showing an example of casting it might also be helpful to show an example of casting to a non-core object (ie: a user-defined class).

File: lib/optparse.rb (hunk 6 above)
Comment (joelmccracken) [inline lib/optparse.rb:281]: Yeah, I agree. I think these would make more sense in their own examples. I'll look into that.

**Comment by joelmccracken** [comment]
sure, I'll look into that.

File: lib/optparse.rb (hunk 2 above)
Comment (joelmccracken) [inline lib/optparse.rb]: what do you mean by this? The change is bad?

File: lib/optparse.rb (hunk 5 above)
Comment (joelmccracken) [inline lib/optparse.rb]: Sure

**Comment by joelmccracken** [comment]
So, one of the things I have gotten hung up on is that I would like to document the completion facilities, but I'm not sure how they work, and cant seem to find any resources on it. Should I file an issue on the ruby bugtracker as it says in these docs?

**Comment by zzak** [comment]
I think this example is getting better, but we seem to be passing a lot of shared knowledge around. Many methods repeating the same parameters: `foo(parser, options)`.

To me this could be better, we shouldn't have to pass around these arguments over and over again.

What do you think?

**Comment by joelmccracken** [comment]
sure -- I'll add them as instance variables.

**Comment by joelmccracken** [comment]
So, I have a proposal. I would like to have two separate kinds of examples:

The first kind of example is a very simple example that displays all the
features. There will be many of these.

The second kind of example is a complete example that shows how one might
use _some_ of the features in a real application. There will probably only
be one of these.

**Comment by joelmccracken** [comment]
Also, I have noted that it seems like completion is something that there is
code support for, but is not document anywhere and I can't figure out how
it works. I will file a ticket on the ruby issue tracker about it.

**Comment by zzak** [comment]
@joelmccracken Let's get the first one done first, and then we can think about building a more complete example for `optparse`. Thanks so much for continuing to work on this! <3

**Comment by joelmccracken** [comment]
When you say "get the first one done first", are you referring to just
finishing this large example?

I know that in a previous comment you mentioned creating real-ish examples
for the large example. I personally had trouble coming up with anything
that would fit for some examples (e.g. like casting numeric arguments to
floats). That's why I suggested pulling some of them out.

**Comment by zzak** [comment]
@joelmccracken I mean we should start small and try to finish the first example you listed in your last reply. Then we can try to come up with a way to better explain more features in greater detail.

**Comment by joelmccracken** [comment]
:+1:

**Comment by joelmccracken** [comment]
@zzak So this is what I was thinking. Features demonstrated in the larger, complete example can be cannibalized into smaller examples, thus allowing for enough context around examples and having the larger example still make sense as a whole. Make sense?

**Comment by joelmccracken** [comment]
Creating embedded rdoc examples is really tedious. There should be a better way to automate it. Anyway, I think I'm going to hold off -- it seems like lots of these pieces could go upstream.

**Comment by zzak** [comment]
@joelmccracken Thank you! I will try to review this asap! <3 <3 <3

**Comment by zzak** [comment]
Thanks @joelmccracken!! This was merged in ruby/ruby@ae0e6807696b3c7b3372619ec6408dfe83333210

**Pull request closed without merging**